### ISSUE TITLE
[BUG] successfully install from source but No module named "mlx.core"

### ISSUE THREAD

**Issue by sirlaurie** [ml-explore/mlx#454]
**Describe the bug**
mlx was build from source and installed successfully, but python said "No module named 'mlx.core'" and cannot generate stubs
```
Python 3.11.7 (main, Dec  4 2023, 18:10:11) [Clang 15.0.0 (clang-1500.1.0.2.5)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import mlx.core as mlx
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'mlx.core'
``` 

**To Reproduce**
1.  `git clone git@github.com:ml-explore/mlx.git mlx && cd mlx`
2.  `brew install pybind11`
3.  `env CMAKE_BUILD_PARALLEL_LEVEL="" pip install ".[dev]"`
4.  `python setup.py generate_stubs` and ModuleNotFoundError happened in this step

<details>
<summary> install log</summary>

```
…/mlx on  main via △ v3.28.1 via 🐍 v3.11.7

➜  env CMAKE_BUILD_PARALLEL_LEVEL="" pip install ".[dev]"
Processing /Users/user/Downloads/mlx
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Installing backend dependencies ... done
  Preparing metadata (pyproject.toml) ... done
Requirement already satisfied: pre-commit in /opt/homebrew/lib/python3.11/site-packages (from mlx==0.0.9.dev2024114+6e81c3e) (3.6.0)
Collecting pybind11-stubgen (from mlx==0.0.9.dev2024114+6e81c3e)
  Using cached pybind11_stubgen-2.4.2-py3-none-any.whl.metadata (1.7 kB)
Requirement already satisfied: cfgv>=2.0.0 in /opt/homebrew/lib/python3.11/site-packages (from pre-commit->mlx==0.0.9.dev2024114+6e81c3e) (3.4.0)
Requirement already satisfied: identify>=1.0.0 in /opt/homebrew/lib/python3.11/site-packages (from pre-commit->mlx==0.0.9.dev2024114+6e81c3e) (2.5.33)
Requirement already satisfied: nodeenv>=0.11.1 in /opt/homebrew/lib/python3.11/site-packages (from pre-commit->mlx==0.0.9.dev2024114+6e81c3e) (1.8.0)
Requirement already satisfied: pyyaml>=5.1 in /opt/homebrew/lib/python3.11/site-packages (from pre-commit->mlx==0.0.9.dev2024114+6e81c3e) (6.0.1)
Requirement already satisfied: virtualenv>=20.10.0 in /opt/homebrew/lib/python3.11/site-packages (from pre-commit->mlx==0.0.9.dev2024114+6e81c3e) (20.25.0)
Requirement already satisfied: setuptools in /opt/homebrew/lib/python3.11/site-packages (from nodeenv>=0.11.1->pre-commit->mlx==0.0.9.dev2024114+6e81c3e) (69.0.2)
Requirement already satisfied: distlib<1,>=0.3.7 in /opt/homebrew/lib/python3.11/site-packages (from virtualenv>=20.10.0->pre-commit->mlx==0.0.9.dev2024114+6e81c3e) (0.3.8)
Requirement already satisfied: filelock<4,>=3.12.2 in /opt/homebrew/lib/python3.11/site-packages (from virtualenv>=20.10.0->pre-commit->mlx==0.0.9.dev2024114+6e81c3e) (3.13.1)
Requirement already satisfied: platformdirs<5,>=3.9.1 in /opt/homebrew/lib/python3.11/site-packages (from virtualenv>=20.10.0->pre-commit->mlx==0.0.9.dev2024114+6e81c3e) (4.1.0)
Using cached pybind11_stubgen-2.4.2-py3-none-any.whl (29 kB)
Building wheels for collected packages: mlx
  Building wheel for mlx (pyproject.toml) ... done
  Created wheel for mlx: filename=mlx-0.0.9.dev2024114+6e81c3e-cp311-cp311-macosx_14_0_arm64.whl size=11431839 sha256=18b83025e73176856c0f1c2c14766a91e4e3fcecc8acc3f84aad51bd692e4683
  Stored in directory: /private/var/folders/tw/7b3zd1w561z2m25dn0pz776h0000gn/T/pip-ephem-wheel-cache-cpt9tdq_/wheels/62/71/3d/0603f5068744a7bae1e68d1d256899ae698ad800182b546649
Successfully built mlx
Installing collected packages: pybind11-stubgen, mlx
Successfully installed mlx-0.0.9.dev2024114+6e81c3e pybind11-stubgen-2.4.2

```

</details>

**Expected behavior**
could import without error

**Desktop (please complete the following information):**
 - Apple M1 Max
 - OS Version: MacOS 14.2.1
 - Python Version: 3.11
 - Xcode Version: 15.2

**Comment by awni** [comment]
Could you check `pip --version` and make sure it matches `python -m pip --version`? It's possible your pip is installing somewhere else in which case you can try:

`env CMAKE_BUILD_PARALLEL_LEVEL="" python -m pip install -e .`

**Comment by sirlaurie** [comment]
it matched.

```shell
➜  pip --version
pip 23.3.2 from /opt/homebrew/lib/python3.11/site-packages/pip (python 3.11)

~
➜  python3 -m pip --version
pip 23.3.2 from /opt/homebrew/lib/python3.11/site-packages/pip (python 3.11)
```

**Comment by sirlaurie** [comment]
and here is mlx package directory

```shell
➜  ls /opt/homebrew/lib/python3.11/site-packages/mlx/
 __pycache__                  extension.py   nn              utils.py
 _reprlib_fix.py              include        optimizers.py
 core.cpython-312-darwin.so   lib           share
```

there's no file or directory name `core`. is that normal?

**Comment by awni** [comment]
The code file is there `core.cpython-312-darwin.so`. From inside that directory you should be able to `import core`.

It seems like that path isn't getting added to your python path (`/opt/homebrew/lib/python3.11/site-packages/mlx/`). Maybe try making a conda environment and using it to install ....

**Comment by sirlaurie** [comment]
from inside `mlx` directory, it still cannot `import core`

```shell
…/homebrew/lib/python3.11/site-packages/mlx on stable via 🐍 v3.11.7
➜  python3
Python 3.11.7 (main, Dec  4 2023, 18:10:11) [Clang 15.0.0 (clang-1500.1.0.2.5)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import core
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'core'
>>>
```

**Comment by awni** [comment]
I don't know how that happened.. but I would recommend you start from a fresh conda environment and rebuild.

**Comment by sirlaurie** [comment]
> I don't know how that happened.. but I would recommend you start from a fresh conda environment and rebuild.

It indeed strange. every message from installing log point that the version of python which using is python3.11. 

Anyway, I used poetry building a new virtual environment and specified python version to `python3.11`.  In this fresh environment, I build it again. But it got the same problem.

Then I build another new environment and specified python version to `python3.12`. it works!

I completely have no idea how this happened. python3.12 was install automatically by brew just because some brew package depends on it. The daily used python version is python3.11.

now I got a new problem when I run `python setup.py generate_stubs` 

```shell
…/mlx on  main [!] is 📦 v0.1.0 via △ v3.28.1 via 🐍 v3.12.1 (mlx-py3.12) 💫 11s118ms
➜  python setup.py generate_stubs
running generate_stubs
pybind11_stubgen - [  ERROR] In mlx.core.DeviceType.__eq__ : Invalid expression 'mlx::core::Device'
pybind11_stubgen - [  ERROR] In mlx.core._ArrayAt.__init__ : Invalid expression 'mlx::core::array'
pybind11_stubgen - [  ERROR] In mlx.core._ArrayAt.add : Invalid expression 'mlx::core::array'
pybind11_stubgen - [  ERROR] In mlx.core._ArrayAt.divide : Invalid expression 'mlx::core::array'
pybind11_stubgen - [  ERROR] In mlx.core._ArrayAt.maximum : Invalid expression 'mlx::core::array'
pybind11_stubgen - [  ERROR] In mlx.core._ArrayAt.minimum : Invalid expression 'mlx::core::array'
pybind11_stubgen - [  ERROR] In mlx.core._ArrayAt.multiply : Invalid expression 'mlx::core::array'
pybind11_stubgen - [  ERROR] In mlx.core._ArrayAt.subtract : Invalid expression 'mlx::core::array'
pybind11_stubgen - [  ERROR] In mlx.core.grad : Can't find/import 'cpp_function'
pybind11_stubgen - [  ERROR] In mlx.core.value_and_grad : Can't find/import 'cpp_function'
pybind11_stubgen - [  ERROR] In mlx.core.vmap : Can't find/import 'cpp_function'
pybind11_stubgen - [WARNING] Raw C++ types/values were found in signatures extracted from docstrings.
Please check the corresponding sections of pybind11 documentation to avoid common mistakes in binding code:
```

any idea to fix this?

**Comment by awni** [comment]
That's just a minor bug with our stub generation, you can safely ignore it for now. I pushed a fix in #455 .
